tools/splat_ext/pm_sbn.py

```diff
@@ -0,0 +1,886 @@
+    unk_38: int = 0
+    unk_3C: int = 0
+
+    fstring = ">4siii iiii iiii iiii"
+    length = 0x40
+
+    def __iter__(self):
+        return iter(astuple(self))
+
+    def __post_init__(self):
+        assert(self.signature == b"SBN ")
+
+    
+@dataclass
+class SBNFileEntry:
+    offset: int
+    fmt: int
+    unk__05: int = 0
+    size: int = 0
+    
+    fstring = ">iBBh"
+    length = 0x08
+
+    def __iter__(self):
+        return iter(astuple(self))
+
+class SBNFile:
+    signature: bytes
+    size: int
+
+    name: bytes
+    data: bytes
+
+    host_system_name: str
+    # the "size" field of the SBN File entry usually, but not always, matches the actual file size.
+    # since this field is unused (I think), it doesn't actually matter other than for reproducing the original cart
+    # I couldn't figure out the pattern behind these fake sizes (if there even is one), so I'm just hardcoding them.
+    fakesize: int
+    # The file ID of the SBN file - needed to guarantee unique filenames. the filename in the header can't be used because it's not unique
+    ident: int
+    def decode(self, data: bytes, ident: int) -> int:
+        self.signature, self.size, self.name = struct.unpack_from(">4si4s", data)
+        self.data = data[:self.size]
+        self.ident = ident
+        
+        self.fakesize = self.size 
+
+        return self.size
+
+    def write(self, path: Path):
+        with open(path, "wb") as f:
+            f.write(self.data)
+
+    def read(self, path: Path):
+        with open(path, "rb") as f:
+            data = f.read()
+
+        filename = os.path.basename(path)
+
+        ident = int(filename.split("_")[0],16)
+
+        size = self.decode(data, ident)
+
+        assert size == len(data), "File size mismatch"
+
+    def file_name(self) -> str:
+        prefix = f"{self.ident:02X}_"
+        stem = decode_null_terminated_ascii(self.name).rstrip(" ")
+        extension = decode_null_terminated_ascii(self.signature).rstrip(" ").lower()
+        return prefix + stem + "." + extension
+
+    def __str__(self):
+        return "SBNFile(" + self.file_name() + ")"
+
+    def __hash__(self):
+        return hash((self.signature, self.size, self.name, self.data))
+
+
+# Could extend SBNFile as BGM etc in future if we want higher-level formats
+
+
+class INIT:
+    song_entries = List["InitSongEntry"]
+    mseq_entries: List[int]
+    bk_entries: List["BufferEntry"]
+
+    def __init__(self):
+        self.song_entries = []
+        self.mseq_entries = []
+        self.bk_entries = []
+
+    def decode(self, data: bytes) -> int:
+        # see au_load_INIT
+
+        header = INITHeader(*struct.unpack_from(INITHeader.fstring, data))
+
+        data = data[:header.size]
+        # Decode song list
+        song_addr = header.tblOffset
+        song_number = 0
+        while True:
+            song = InitSongEntry(*struct.unpack_from(InitSongEntry.fstring, data, song_addr))
+
+            if(song.bgmFileIndex == 0xffff):
```

Comment:
No parens needed here - `black` should fix this

Reply:
also this should be bgm_file_index

Reply:
This directly corresponds to the InitSongEntry::bgmFileIndex field defined in audio.h. As such I think it makes sense to match the name. Same with all the other fields that match to C struct fields

Reply:
ahh I see. okay, that's a good argument 👍